### PR TITLE
Add security enhancement command when create a cluster or add nodes 

### PR DESCRIPTION
**Pull request by zhouqiu0103**
### What type of PR is this?

/kind feature

### What this PR does / why we need it:
use [CIS kubernetes benchmark](https://www.cisecurity.org/benchmark/kubernetes) to hardening the k8s cluster
create cluster command:
```
./kk create cluster --with-security-enhancement
```
### Which issue(s) this PR fixes: 

Fixes #

### Special notes for reviewers:
```
```

### Does this PR introduced a user-facing change?

```release-note
Add command '--with-security-enhancement' when create k8s cluster or add nodes
```

### Additional documentation, usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/pipelines/create_cluster.go [hunk 1]
```diff
@@ -90,6 +90,7 @@ func NewCreateClusterPipeline(runtime *common.KubeRuntime) error {
 		&kubernetes.ConfigureKubernetesModule{},
 		&filesystem.ChownModule{},
 		&certs.AutoRenewCertsModule{Skip: !runtime.Cluster.Kubernetes.EnableAutoRenewCerts()},
+		&kubernetes.SecurityEnhancementModule{Skip: runtime.Arg.SecurityEnhancement},
```
Comment (24sama) [inline pkg/pipelines/create_cluster.go:93]: if the value of `runtime.Arg.SecurityEnhancement` is true, this module will be skipped.

File: pkg/kubernetes/tasks.go [hunk 2]
```diff
@@ -1071,3 +1072,147 @@ func (c *ConfigureKubernetes) Execute(runtime connector.Runtime) error {
...
+	chmodEtcdCertsCmd := "chmod 600 /etc/ssl/etcd/ssl/*"
+	chmodEtcdDataDirCmd := "chmod 700 /var/lib/etcd"
+	chmodEtcdCmd := "chmod 550 /usr/local/bin/etcd*"
+
+	chownEtcdCertsDirCmd := "chown root:root /etc/ssl/etcd/ssl"
+	chownEtcdCertsCmd := "chown root:root /etc/ssl/etcd/ssl/*"
+	chownEtcdDataDirCmd := "chown etcd:etcd /var/lib/etcd"
+	chownEtcdCmd := "chown root:root /usr/local/bin/etcd*"
+
+	if _, err := runtime.GetRunner().SudoCmd(fmt.Sprintf("%s && %s && %s && %s && %s && %s && %s && %s", chmodEtcdCertsDirCmd, chmodEtcdCertsCmd, chmodEtcdDataDirCmd, chmodEtcdCmd, chownEtcdCertsDirCmd, chownEtcdCertsCmd, chownEtcdDataDirCmd, chownEtcdCmd), true); err != nil {
```
Comment (24sama) [inline pkg/kubernetes/tasks.go:1092]: Suggest:
```go
cmds := []string{chmodEtcdCertsDirCmd, chmodEtcdCertsCmd, chmodEtcdDataDirCmd}
runtime.GetRunner().SudoCmd(strings.Join(arr, " && "), true)
```

File: pkg/kubernetes/tasks.go [hunk 3]
```diff
@@ -1071,3 +1072,147 @@ func (c *ConfigureKubernetes) Execute(runtime connector.Runtime) error {
...
+	}
+
+	chmodNodesCmd := fmt.Sprintf("%s && %s && %s && %s && %s && %s && %s", chmodBinDir, chmodKubeCmd, chmodHelmCmd, chmodCniDir, chmodCniConfigDir, chmodKubeletConfig, chmodCertsRenew)
+	if _, err := runtime.GetRunner().SudoCmd(chmodNodesCmd, true); err != nil {
+		return errors.Wrap(errors.WithStack(err), "Updating permissions failed.")
+	}
+	chownNodesCmd := fmt.Sprintf("%s && %s && %s && %s && %s && %s && %s", chownBinDir, chownKubeCmd, chownHelmCmd, chownCniDir, chownCniConfigDir, chownKubeletConfig, chownCertsRenew)
+	if _, err := runtime.GetRunner().SudoCmd(chownNodesCmd, true); err != nil {
+		return errors.Wrap(errors.WithStack(err), "Updating permissions failed.")
+	}
```
Comment (24sama) [inline pkg/kubernetes/tasks.go:1156]: The same as above.

File: pkg/kubernetes/tasks.go [hunk 4]
```diff
@@ -1071,3 +1072,147 @@ func (c *ConfigureKubernetes) Execute(runtime connector.Runtime) error {
...
+
+	chmodNodesCmd := fmt.Sprintf("%s && %s && %s && %s && %s && %s", chmodBinDir, chmodKubeCmd, chmodHelmCmd, chmodCniDir, chmodCniConfigDir, chmodKubeletConfig)
+	if _, err := runtime.GetRunner().SudoCmd(chmodNodesCmd, true); err != nil {
+		return errors.Wrap(errors.WithStack(err), "Updating permissions failed.")
+	}
+	chownNodesCmd := fmt.Sprintf("%s && %s && %s && %s && %s && %s", chownBinDir, chownKubeCmd, chownHelmCmd, chownCniDir, chownCniConfigDir, chownKubeletConfig)
+	if _, err := runtime.GetRunner().SudoCmd(chownNodesCmd, true); err != nil {
+		return errors.Wrap(errors.WithStack(err), "Updating permissions failed.")
+	}
```
Comment (24sama) [inline pkg/kubernetes/tasks.go:1215]: The same as above.

File: pkg/kubernetes/module.go [hunk 5]
```diff
@@ -598,3 +604,43 @@ func (c *ConfigureKubernetesModule) Init() {
 		configure,
 	}
 }
+
+type SecurityEnhancementModule struct {
+	common.KubeModule
+	Skip bool
+}
+
```
Comment (24sama) [inline pkg/kubernetes/module.go:612]: Missing override IsSkip() method

**Review by 24sama — COMMENTED** [review]
There are some comments, PTAL.

**Comment by 24sama** [comment]
And others, could you please paste some more detailed links or references here?  That can help the users understand why these configurations are needed.

File: pkg/kubernetes/module.go [hunk 6]
```diff
@@ -598,3 +604,47 @@ func (c *ConfigureKubernetesModule) Init() {
...
+		Desc:     "Security enhancement for kubernetes",
+		Hosts:    s.Runtime.GetHostsByRole(common.Master),
+		Action:   new(MasterSecurityEnhancemenAction),
+		Parallel: true,
+	}
+
+	nodesSecurityEnhancement := &task.RemoteTask{
+		Name:     "K8sSecurityEnhancementTask",
+		Desc:     "Security enhancement for kubernetes",
+		Hosts:    s.Runtime.GetHostsByRole(common.K8s),
```
Comment (24sama) [inline pkg/kubernetes/module.go:640]: I found that this `action` is a subset of the `MasterSecurityEnhancemenAction`, so maybe I guess it should be `s.Runtime.GetHostsByRole(common.Worker)`

Comment (zhouqiu0103) [reply pkg/kubernetes/module.go:640]: I think so

**Comment by zhouqiu0103** [comment]
> And others, could you please paste some more detailed links or references here? That can help the users understand why these configurations are needed.

The **CIS kubernetes benchmark** provides recommendations for hardening in five main directions. These are `Control Plane`, `etcd`, `Worker Nodes` and `Kubernetes policies`.Considering that `Kubernetes policies` are mainly about things that users should be aware of while using k8s clusters (e.g. Pod Security Standards), I had ignored this section.Details are shown below.

**Control Plane**
1. Restrict access and ownership for control-plane nodes' configuration files and their folders (such as kube-apiserver.yaml, admin.conf, etc.).The files and folders involved are shown below.
    ```
    - /etc/kubernetes
    - /etc/kubernetes/*             
    - /etc/kubernetes/manifests
    - /etc/kubernetes/manifests/*    
    - /etc/kubernetes/pki
    - /etc/kubernetes/pki/*
    - /etc/cni/net.d
    - /usr/local/bin
    - /usr/local/bin/kube*
    - /usr/local/bin/helm
    - /opt/cni/bin
    - /opt/cni/bin/*
    - /etc/systemd/system/k8s-certs-renew*
    ```
2. API server configuration flags. Details are shown below.
    ```
    - --enable-admission-plugins=AlwaysPullImages,ServiceAccount,NamespaceLifecycle,NodeRestriction,LimitRanger,ResourceQuota,MutatingAdmissionWebhook,ValidatingAdmissionWebhook,PodNodeSelector,PodSecurity
    - --profiling=false
    - --request-timeout=120s
    - --service-account-lookup=true
    - --tls-cipher-suites=TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305
    - --tls-min-version=VersionTLS12
    ```
3. Controller Manager configuration flags. Details are shown below.
    ```
    - --bind-address=127.0.0.1
    - --profiling=false
    - --terminated-pod-gc-threshold=50
    - --use-service-account-credentials=true
    ```
4. Scheduler configuration flags. Details are shown below.
    ```
    - --bind-address=127.0.0.1
    - --profiling=false
    ```

**etcd**

Restrict access and ownership for etcd-related certificates, data files, and binaries.The files and folders involved are shown below.
```
- /etc/ssl/etcd/ssl
- /etc/ssl/etcd/ssl/*
- /var/lib/etcd
- /usr/local/bin/etcd*
```

**Worker Nodes**
1. Restrict access and ownership for worker nodes' configuration files and their folders.The files and folders involved are shown below.
    ```
    - /var/lib/kubelet/config.yaml
    - /etc/systemd/system/kubelet.service*
    ```
2. Kubelet configuration flags. Details are shown below.
    ```
    - --event-qps=1
    - --make-iptables-util-chains=true
    - --protect-kernel-defaults=true
    - --read-only-port=0
    - --streaming-connection-idle-timeout=5m
    - --tls-cipher-suites=TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305
    ```
In addition to the above configurations, there are several configurations such as `--enable-admission-plugins=EventRateLimit`,`--audit-log-path`,`--encryption-provider-config` that require a configuration file to be provided and have not been integrated yet.The specific impact of these configurations and the reasons for setting them up can be found in the official [documentation](https://www.cisecurity.org/benchmark/kubernetes) for the CIS kubernetes benchmark.

File: cmd/ctl/create/cluster.go [hunk 7]
```diff
@@ -135,6 +138,7 @@ func (o *CreateClusterOptions) AddFlags(cmd *cobra.Command) {
 	cmd.Flags().BoolVarP(&o.EnableKubeSphere, "with-kubesphere", "", false, fmt.Sprintf("Deploy a specific version of kubesphere (default %s)", kubesphere.Latest().Version))
 	cmd.Flags().BoolVarP(&o.SkipPullImages, "skip-pull-images", "", false, "Skip pre pull images")
 	cmd.Flags().BoolVarP(&o.SkipPushImages, "skip-push-images", "", false, "Skip pre push images")
+	cmd.Flags().BoolVarP(&o.SecurityEnhancement, "with-security-enhancement", "", false, "Security enhancement")
```
Comment (LinuxSuRen) [inline cmd/ctl/create/cluster.go:141]: I'm wondering why the default value is `false` instead of `true`.

**Pull request merged**